File: services/photographer-service/controllers/photographerController.js
const { Storage } = require('@google-cloud/storage');
const admin = require('firebase-admin');
require('dotenv').config();
const multer = require('multer');
const { v4: uuidv4 } = require("uuid");
const {calculateDistance} = require("../../../helpers/calculateDistance")
const { getCoordinateFromAdress } = require('../../../helpers/getCoordinatesFromAdress') 
const { uploadSingleImage, uploadMultipleImages } = require('../../../helpers/imageUpload/imageUpload')


// Initialize Firestore and Storage
const db = admin.firestore();
const storage = new Storage();
const bucket = admin.storage().bucket()
const upload = multer({ storage: storage }); // Use memory storage for multer

// Function to handle the photographer request
exports.requestPhotographer = async (req, res) => {
    console.log("New request for photographer");

    // Destructure the required fields from req.body
    const { name, surname, description, age, state, operationLocation, phoneNumber, userId, isProfessional, price, profile_picture } = req.body;
    console.log("Destructured fields:", { name, surname, description, age, state, operationLocation, phoneNumber, userId, isProfessional, price, profile_picture });

    // Add profile picture 
    let work = []; // Array to hold image URLs
    const sentFiles = req.files; // Access uploaded files from req.files
    console.log("Received files:", sentFiles);

    // Validate input
    if (!name || !surname || !description || !age || !state || !operationLocation || !phoneNumber) {
        console.log("Validation failed: All fields are required.");
        return res.status(400).json({ message: 'All fields are required.' });
    }
    console.log("Step 1 passed: All required fields are present");

    try {
        // Check if files are present
        if (!sentFiles || sentFiles.length === 0) {
            console.log("Validation failed: No files uploaded.");
            return res.status(400).json({ message: 'No files uploaded.' });
        }
        console.log("Step 2 passed: Files are present");
        console.log("Number of files uploaded:", sentFiles.length);

        // Extract the file paths or URLs from sentFiles
        const imageUrls = await uploadMultipleImages(sentFiles); // Make sure this function handles an array of files
        work = imageUrls;
        console.log("Image URLs uploaded:", work);

        // Save photographer data to Firestore
        const newPhotographerRef = db.collection('photographers').doc(); // Create a new document with a generated ID
        const newPhotographer = {
            name,
            surname,
            description,
            age,
            state,
            work,
            operationLocation,
            phoneNumber,
            userId,
            isProfessional,
            price,
            profile_picture : "https://i.sstatic.net/l60Hf.png",
            status: 0, // Default status
            note : 5,
        };
        console.log("New photographer data:", newPhotographer);

        await newPhotographerRef.set(newPhotographer);
        console.log("Photographer added to Firestore with ID:", newPhotographerRef.id);
        res.status(201).json({ message: 'Photographer added successfully!', photographerId: newPhotographerRef.id });
    } catch (error) {
        console.error("Error occurred while processing the request:", error);
        res.status(500).json({ message: 'Internal server error.' });
    }
}

// Endpoint to get nearest photographers
exports.getNearestPhotographers = async (req, res) => {
    const { userPosition, type, expertise } = req.body;

    if (!type || !expertise) {
        console.log("treating without filters")
    }
    try {
        // gets all the photographers
        const photographersRef = db.collection('photographers');
        const snapshot = await photographersRef.get();

        const photographers = [];
        snapshot.forEach(doc => {
        const data = doc.data();
        photographers.push({
            id: doc.id,
            ...data
        });
        });
        
        // Get all the photographers' coordinates and calculate distances concurrently
        let nearPhotographers = [];

        // Create an array of promises for fetching coordinates and calculating distances
        const photographerPromises = photographers.map(async (photographer) => {
            const photographerCoordinate = await getCoordinateFromAdress(photographer.operationLocation);
            const distance = await calculateDistance(userPosition, photographerCoordinate);
            
            // Check if the photographer is within the desired distance
            if (distance < 100) {
                return photographer; // Return the photographer if they are near
            } else {
                return null; // Return null if they are too far
            }
        });

        // Wait for all promises to resolve
        const results = await Promise.all(photographerPromises);

        // Filter out null values (i.e., photographers that are too far)
        nearPhotographers = results.filter((photographer) => photographer !== null);

        console.log(nearPhotographers);
        
        res.status(200).json(nearPhotographers);

    } catch (error) {
        console.error(error);
        res.status(500).json({ error: 'An error occurred while fetching photographers' });
    }
}

exports.isPhotographer = async (req, res) => {
    try {
        const { userId } = req.body; // Extracting userId from the request body

        // Ensure that userId is provided and is a string
        if (!userId || typeof userId !== 'string') {
            return res.status(400).json({ message: 'Invalid or missing userId' });
        }

        const photographersRef = db.collection('photographers');
        const snapshot = await photographersRef.where('userId', '==', userId).get();

        if (!snapshot.empty) {
            // User exists as a photographer, return the photographerId
            const photographer = snapshot.docs[0].data();
            const photographerId = snapshot.docs[0].id; // Assuming the document ID is the photographerId
            return res.status(200).json({ isPhotographer : photographerId });
        } else {
            // User does not exist as a photographer
            return res.status(200).json({ isPhotographer: false });
        }
    } catch (error) {
        console.error('Error checking photographer status:', error);
        return res.status(500).json({ message: 'Could not check photographer status' });
    }
};

exports.getPhotographerById = async (req, res) => {
    try {
        const { photographerId } = req.body; // Extracting photographerId from the request body

        if (!photographerId || typeof photographerId !== 'string') {
            console.log("Invalid or missing photographerId");
            return res.status(400).json({ message: 'Invalid or missing photographerId' });
        }

        const photographersRef = db.collection('photographers');
        const snapshot = await photographersRef.doc(photographerId).get(); // Use .doc() to directly access the document by ID

        if (snapshot.exists) {
            const photographer = snapshot.data();
            return res.status(200).json({ photographer });
        } else {
            console.log(`No photographer found for ID: ${photographerId}`);
            return res.status(404).json({ message: 'No photographer found' });
        }
    } catch (error) {
        console.error('Error retrieving photographer:', error);
        return res.status(500).json({ message: 'Could not retrieve photographer data' });
    }
};

exports.loadDashboard = async (req, res) => {
    try {
      const { photographerId } = req.body; // Get photographerId from request body
  
      if (!photographerId) {
        return res.status(400).json({ error: "photographerId is required" });
      }
  
      // Query the Firestore 'propositions' collection where p2_id matches photographerId
      const querySnapshot = await db
        .collection('propositions')
        .where('p2_id', '==', photographerId)
        .get();
  
      if (querySnapshot.empty) {
        console.log("not found")

        return res.status(400).json({ message: "No propositions found for this photographer" });
      }
  
      // Prepare the results
      const propositions = querySnapshot.docs.map((doc) => ({
        id: doc.id,
        ...doc.data(),
      }));
  
      return res.status(200).json(propositions); // Return the list of propositions
    } catch (error) {
      console.error("Error retrieving propositions: ", error);
      return res.status(500).json({ error: "Something went wrong" });
    }
  };

  exports.updateOperationLocation = async (req, res) => {
    const { photographerId, newValue } = req.body;
  
    try {
      const photographerRef = db.collection('photographers').doc(photographerId);
      await photographerRef.update({
        operationLocation: newValue
      });
  
      res.status(200).send({ message: 'Operation Location updated successfully' });
    } catch (error) {
      res.status(500).send({ message: 'Error updating Operation Location', error: error.message });
    }
  };
  
  exports.updateState = async (req, res) => {
    console.log("received")
    const { photographerId, newValue } = req.body;
    console.log(newValue)
    try {
      const photographerRef = db.collection('photographers').doc(photographerId);
      await photographerRef.update({
        state: newValue
      });
  
      res.status(200).send({ message: 'State updated successfully' });
    } catch (error) {
        console.log("no photographer found")
      res.status(500).send({ message: 'Error updating State', error: error.message });
    }
  };
  exports.updateDescription = async (req, res) => {
    const { photographerId, newValue } = req.body;
  
    try {
      const photographerRef = db.collection('photographers').doc(photographerId);
      await photographerRef.update({
        description: newValue
      });
  
      res.status(200).send({ message: 'Description updated successfully' });
    } catch (error) {
      res.status(500).send({ message: 'Error updating Description', error: error.message });
    }
  };
  exports.updatePrice = async (req, res) => {
    const { photographerId, newValue } = req.body;
  
    try {
      const photographerRef = db.collection('photographers').doc(photographerId);
      await photographerRef.update({
        price: newValue
      });
  
      res.status(200).send({ message: 'Price updated successfully' });
    } catch (error) {
      res.status(500).send({ message: 'Error updating Price', error: error.message });
    }
  };
  

File: services/photographer-service/controllers/photographerController.py
from flask import Flask, request, jsonify
import os
import firebase_admin
from firebase_admin import firestore, storage
from dotenv import load_dotenv
import uuid
from helpers.calculate_distance import calculate_distance
from helpers.get_coordinates_from_address import get_coordinate_from_address
from helpers.image_upload import upload_single_image, upload_multiple_images

# Load environment variables from a .env file
load_dotenv()

# Initialize Firestore and Storage
firebase_admin.initialize_app()
db = firestore.client()
bucket = storage.bucket()

app = Flask(__name__)

@app.route('/requestPhotographer', methods=['POST'])
async def request_photographer():
    print("New request for photographer")

    data = request.form
    files = request.files.getlist('files')

    required_fields = ['name', 'surname', 'description', 'age', 'state', 'operationLocation', 'phoneNumber', 'userId', 'isProfessional', 'price', 'profile_picture']

    for field in required_fields:
        if field not in data:
            return jsonify({'message': 'All fields are required.'}), 400

    if not files:
        return jsonify({'message': 'No files uploaded.'}), 400

    try:
        image_urls = await upload_multiple_images(files)
        work = image_urls

        new_photographer_ref = db.collection('photographers').document()
        new_photographer = {
            'name': data['name'],
            'surname': data['surname'],
            'description': data['description'],
            'age': data['age'],
            'state': data['state'],
            'work': work,
            'operationLocation': data['operationLocation'],
            'phoneNumber': data['phoneNumber'],
            'userId': data['userId'],
            'isProfessional': data['isProfessional'],
            'price': data['price'],
            'profile_picture': "https://i.sstatic.net/l60Hf.png",
            'status': 0,
            'note': 5,
        }

        await new_photographer_ref.set(new_photographer)
        return jsonify({'message': 'Photographer added successfully!', 'photographerId': new_photographer_ref.id}), 201

    except Exception as error:
        print("Error occurred while processing the request:", error)
        return jsonify({'message': 'Internal server error.'}), 500

@app.route('/getNearestPhotographers', methods=['POST'])
async def get_nearest_photographers():
    data = request.json
    user_position = data.get('userPosition')
    type_filter = data.get('type')
    expertise_filter = data.get('expertise')

    try:
        photographers_ref = db.collection('photographers')
        snapshot = await photographers_ref.get()

        photographers = [{'id': doc.id, **doc.to_dict()} for doc in snapshot]

        near_photographers = []

        for photographer in photographers:
            photographer_coordinate = await get_coordinate_from_address(photographer['operationLocation'])
            distance = await calculate_distance(user_position, photographer_coordinate)

            if distance < 100:
                near_photographers.append(photographer)

        return jsonify(near_photographers), 200

    except Exception as error:
        print("Error fetching photographers:", error)
        return jsonify({'error': 'An error occurred while fetching photographers'}), 500

@app.route('/isPhotographer', methods=['POST'])
async def is_photographer():
    data = request.json
    user_id = data.get('userId')

    if not user_id or not isinstance(user_id, str):
        return jsonify({'message': 'Invalid or missing userId'}), 400

    try:
        photographers_ref = db.collection('photographers')
        snapshot = await photographers_ref.where('userId', '==', user_id).get()

        if snapshot:
            photographer_id = snapshot[0].id
            return jsonify({'isPhotographer': photographer_id}), 200
        else:
            return jsonify({'isPhotographer': False}), 200

    except Exception as error:
        print("Error checking photographer status:", error)
        return jsonify({'message': 'Could not check photographer status'}), 500

@app.route('/getPhotographerById', methods=['POST'])
async def get_photographer_by_id():
    data = request.json
    photographer_id = data.get('photographerId')

    if not photographer_id or not isinstance(photographer_id, str):
        return jsonify({'message': 'Invalid or missing photographerId'}), 400

    try:
        photographer_ref = db.collection('photographers').document(photographer_id)
        snapshot = await photographer_ref.get()

        if snapshot.exists:
            photographer = snapshot.to_dict()
            return jsonify({'photographer': photographer}), 200
        else:
            return jsonify({'message': 'No photographer found'}), 404

    except Exception as error:
        print("Error retrieving photographer:", error)
        return jsonify({'message': 'Could not retrieve photographer data'}), 500

@app.route('/loadDashboard', methods=['POST'])
async def load_dashboard():
    data = request.json
    photographer_id = data.get('photographerId')

    if not photographer_id:
        return jsonify({'error': 'photographerId is required'}), 400

    try:
        query_snapshot = await db.collection('propositions').where('p2_id', '==', photographer_id).get()

        if not query_snapshot:
            return jsonify({'message': 'No propositions found for this photographer'}), 400

        projects = [{'id': doc.id, **doc.to_dict()} for doc in query_snapshot]

        enriched_projects = []
        for project in projects:
            user_doc = await db.collection('Users').document(project['p1_id']).get()
            if user_doc.exists:
                enriched_projects.append({**project, 'client': user_doc.to_dict()})
            else:
                enriched_projects.append(project)

        return jsonify(enriched_projects), 200

    except Exception as error:
        print("Error retrieving propositions:", error)
        return jsonify({'error': 'Something went wrong'}), 500

@app.route('/updateOperationLocation', methods=['POST'])
async def update_operation_location():
    data = request.json
    photographer_id = data.get('photographerId')
    new_value = data.get('newValue')

    try:
        photographer_ref = db.collection('photographers').document(photographer_id)
        await photographer_ref.update({'operationLocation': new_value})
        return jsonify({'message': 'Operation Location updated successfully'}), 200

    except Exception as error:
        return jsonify({'message': 'Error updating Operation Location', 'error': str(error)}), 500

@app.route('/updateState', methods=['POST'])
async def update_state():
    data = request.json
    photographer_id = data.get('photographerId')
    new_value = data.get('newValue')

    try:
        photographer_ref = db.collection('photographers').document(photographer_id)
        await photographer_ref.update({'state': new_value})
        return jsonify({'message': 'State updated successfully'}), 200

    except Exception as error:
        return jsonify({'message': 'Error updating State', 'error': str(error)}), 500

@app.route('/updateDescription', methods=['POST'])
async def update_description():
    data = request.json
    photographer_id = data.get('photographerId')
    new_value = data.get('newValue')

    try:
        photographer_ref = db.collection('photographers').document(photographer_id)
        await photographer_ref.update({'description': new_value})
        return jsonify({'message': 'Description updated successfully'}), 200

    except Exception as error:
        return jsonify({'message': 'Error updating Description', 'error': str(error)}), 500

@app.route('/updatePrice', methods=['POST'])
async def update_price():
    data = request.json
    photographer_id = data.get('photographerId')
    new_value = data.get('newValue')

    try:
        photographer_ref = db.collection('photographers').document(photographer_id)
        await photographer_ref.update({'price': new_value})
        return jsonify({'message': 'Price updated successfully'}), 200

    except Exception as error:
        return jsonify({'message': 'Error updating Price', 'error': str(error)}), 500

@app.route('/addImageToWork', methods=['POST'])
async def add_image_to_work():
    data = request.form
    photographer_id = data.get('photographerId')
    file = request.files.get('file')

    try:
        image_url = await upload_single_image(file)
        photographer_ref = db.collection('photographers').document(photographer_id)
        photographer_snapshot = await photographer_ref.get()

        if photographer_snapshot.exists:
            photographer_data = photographer_snapshot.to_dict()
            current_work = photographer_data.get('work', [])
            current_work.append(image_url)
            await photographer_ref.update({'work': current_work})
            return jsonify({'message': 'Image added to work list successfully', 'imageUrl': image_url}), 200
        else:
            return jsonify({'message': 'No photographer found with the provided ID'}), 404

    except Exception as error:
        return jsonify({'message': str(error)}), 500

@app.route('/removeImageFromWork', methods=['POST'])
async def remove_image_from_work():
    data = request.json
    photographer_id = data.get('photographerId')
    image_url = data.get('imageUrl')

    try:
        photographer_ref = db.collection('photographers').document(photographer_id)
        photographer_snapshot = await photographer_ref.get()

        if photographer_snapshot.exists:
            photographer_data = photographer_snapshot.to_dict()
            current_work = photographer_data.get('work', [])
            current_work = [image for image in current_work if image != image_url]
            await photographer_ref.update({'work': current_work})
            return jsonify({'message': 'Image removed from work list successfully'}), 200
        else:
            return jsonify({'message': 'No photographer found with the provided ID'}), 404

    except Exception as error:
        return jsonify({'message': str(error)}), 500

if __name__ == '__main__':
    app.run(debug=True)
